### PR TITLE
Create or find Synapse object instead of update

### LINKED ISSUE #7: is a global 'create or update' what we would want?
Think about if one would want to decide to create some things but update others. I don't know the answer but would influence the design.

### LINKED ISSUE #8: make create functions more generic wrappers
Couldn't kwargs just be passed to the object constructors? This would enable the args to come more directly from the JSON that we'll eventually consume to run these functions.

### LINKED ISSUE #9: `create_or_update` param name is confusing
Not quite clear by the name, because nothing is 'updated' in the current code - it only returns existing objects if they exist. Maybe it's because the param is performing two duties - one to decide to try and get an existing object, and another to pass to the `store` function to do the create or update?

### LINKED ISSUE #14: Describe what 409 error is
Synapse conflict error

### PR DESCRIPTION
**Pull request by thomasyu888**
Fixes #7 
Fixes #8 
Fixes #9 
Fixes #10
Fixes #14

### CODE REVIEW

File: synapseformation/create.py [hunk 1]
```diff
@@ -1,198 +1,207 @@
...
-    def __init__(self, syn: 'Synapse', create_or_update: bool = False,
+    def __init__(self, syn: 'Synapse', only_create: bool = True,
                  logger: Logger = None):
         """
         Args:
             syn: Synapse connection
-            create_or_update: Default is False, which means resources can
-                              only be created and not updated if resource
-                              already exists.
+            only_create: Only create entities.  Default is True, which
```
Comment (kdaily) [inline synapseformation/create.py]: extra space after period

File: synapseformation/create.py [hunk 2]
```diff
@@ -1,198 +1,207 @@
...
-            syn: Synapse connection
-            project_name: Name of project
+            entity: synapseclient.Entity type
+
+        Returns:
+            synapseclient.Entity
+        """
+        try:
+            entity = self.syn.store(entity, createOrUpdate=False)
+        except SynapseHTTPError:
```
Comment (kdaily) [inline synapseformation/create.py]: check the error response or code

File: synapseformation/create.py [hunk 3]
```diff
@@ -1,198 +1,207 @@
...
+            if self.only_create:
+                raise ValueError("only_create is set to True.")
```
Comment (kdaily) [inline synapseformation/create.py]: print contents of the original exception also

File: synapseformation/create.py [hunk 4]
```diff
@@ -1,198 +1,207 @@
...
+            body = json.dumps({"parentId": entity.properties.get("parentId", None),  # pylint: disable=line-too-long
+                               "entityName": entity.name})
+            entity_obj = self.syn.restPOST("/entity/child", body=body)
+            entity_tmp = self.syn.get(entity_obj['id'], downloadFile=False)
+            assert entity.properties.concreteType == entity_tmp.properties.concreteType, "Different types."  # pylint: disable=line-too-long
```
Comment (kdaily) [inline synapseformation/create.py]: shouldn't be an assertion, should be a test and raise an error (ValueError)

Comment (kdaily) [reply synapseformation/create.py]: TypeError?

File: synapseformation/create.py [hunk 5]
```diff
@@ -1,198 +1,207 @@
...
-        self.logger.info('{} Folder {} ({})'.format(self._update_str,
-                                                    folder_ent.name,
-                                                    folder_ent.id))
-        return folder_ent
+        try:
+            team = Team(name=name, *args, **kwargs)
+            team = self.syn.store(team, createOrUpdate=False)
+        except SynapseHTTPError:
+            if self.only_create:
+                raise ValueError("only_create is set to True.")
```
Comment (kdaily) [inline synapseformation/create.py]: get error message from HTTP error too

**Review by kdaily — CHANGES_REQUESTED** [review]
only use `**kwargs`, not `*args`

File: synapseformation/create.py [hunk 6]
```diff
@@ -1,198 +1,216 @@
 """Convenience functions to create Synapse entities"""
+from functools import partial
```
Comment (thomasyu888) [inline synapseformation/create.py]:
```suggestion
```

File: synapseformation/create.py [hunk 7]
```diff
@@ -3,196 +3,215 @@
...
+            obj = self.syn.getTeam(obj)
+        elif isinstance(obj, Wiki):
+            obj = self.syn.getWiki(obj)
+        elif isinstance(obj, Evaluation):
+            obj = self.syn.getEvaluation(obj)
+        else:
+            raise ValueError(f"{obj} not recognized")
+        return obj
+
+    def _find_by_obj_or_create(self, obj: 'Object') -> 'Object':
```
Comment (thomasyu888) [inline synapseformation/create.py]: Use Union

File: tests/test_create.py [hunk 8]
```diff
@@ -7,107 +7,231 @@
...
+        assert created_ent == returned
+
+
+#@patch.object(SYN, "store")
+def test__find_by_obj_or_create__onlycreate_raise():
+    """Tests only create flag raises error when entity exists"""
+    entity = synapseclient.Entity(name=str(uuid.uuid1()))
+    returned = synapseclient.Entity(name=str(uuid.uuid1()))
+    # Mock SynapseHTTPError with 409 response
+    mocked_409 = SynapseHTTPError("foo", response=Mock(status_code=409))
```
Comment (kdaily) [inline tests/test_create.py:42]: note what a 409 error is (a conflict) - is this documented in our REST API?

**Review by kdaily — APPROVED** [review]

**Pull request merged**